lib/actions-oclif/login.ts

```diff
@@ -0,0 +1,188 @@
+
+		- Credentials: using email/password and 2FA.
+
+		- Token: using a session token or API key from the preferences page.
+`;
+	public static examples = [
+		'$ balena login',
+		'$ balena login --web',
+		'$ balena login --token "..."',
+		'$ balena login --credentials',
+		'$ balena login --credentials --email johndoe@gmail.com --password secret',
+	];
+
+	public static args = [
+		{
+			// Capitano allowed -t to be type boolean|string, which oclif does not.
+			// So -t is now bool, and we check first arg for token content.
+			name: 'token',
+			hidden: true,
+		},
+	];
+
+	public static usage = 'login';
+
+	public static flags: flags.Input<FlagsDef> = {
+		web: flags.boolean({
+			char: 'w',
+			description: 'web-based login',
+		}),
+		token: flags.boolean({
+			char: 't',
+			description: 'session token or API key',
+		}),
+		credentials: flags.boolean({
+			char: 'c',
+			description: 'credential-based login',
+		}),
+		email: flags.string({
+			char: 'e',
+			description: 'email',
+			exclusive: ['user'],
+			dependsOn: ['credentials'],
+		}),
+		// Capitano version of this command had a second alias for email, 'u'.
+		// Using an oclif hidden flag to support the same behaviour.
+		user: flags.string({
+			char: 'u',
+			hidden: true,
+			exclusive: ['email'],
+			dependsOn: ['credentials'],
+		}),
+		password: flags.string({
+			char: 'p',
+			description: 'password',
+			dependsOn: ['credentials'],
+		}),
+		help: cf.help,
+	};
+
+	public static primary = true;
+
+	public async run() {
+		const { flags: options, args: params } = this.parse<FlagsDef, ArgsDef>(
+			LoginCmd,
+		);
+
+		const balena = getBalenaSdk();
+		const messages = await import('../utils/messages');
+		const balenaUrl = await balena.settings.get('balenaUrl');
+
+		// Consolidate user/email options
+		if (options.user != null) {
+			options.email = options.user;
+		}
+
+		console.log(messages.balenaAsciiArt);
+		console.log(`\nLogging in to ${balenaUrl}`);
+		await this.doLogin(options, params.token);
+		const username = await balena.auth.whoami();
+
+		console.info(`Successfully logged in as: ${username}`);
+		console.info(`\
+
+Find out about the available commands by running:
+
+  $ balena help
+
+${messages.reachingOut}`);
+
+		if (options.web) {
+			const { shutdownServer } = await import('../auth');
+			shutdownServer();
+		}
+	}
+
+	async doLogin(loginOptions: FlagsDef, token?: string): Promise<void> {
+		const { exitWithExpectedError } = await import('../errors');
+		const patterns = await import('../utils/patterns');
+		const balena = getBalenaSdk();
+
+		// Token
+		if (loginOptions.token) {
+			if (!token) {
+				const form = await import('resin-cli-form');
+				token = await form.ask({
+					message: 'Session token or API key from the preferences page',
+					name: 'token',
+					type: 'input',
+				});
+			}
+			await balena.auth.loginWithToken(token!);
+			if (!(await balena.auth.whoami())) {
+				exitWithExpectedError('Token authentication failed');
```

Comment:
This could probably be `throw new ExpectedError('Token authentication failed')`

Reply:
Yeah missed that.